### PR TITLE
data_regextract: new function like regextract, returns data container with named captures

### PR DESCRIPTION
**Pull request by tzz**
This proposed function does two things:

* adds a new `regextract` variant that creates a data container rather than "classic" array variables
* the data container is a key-value map where the keys are either the position of the match or, if the capture group was named, the name of the capture group (see the example policy below).  This is extremely nice in combination with Mustache templates, because the resulting data container can be fed directly to them.

In addition:

* the function `regex.c:StringMatchCaptures()` now returns a `Seq` of `Buffer` elements, which is nicer and removes the 4K limitation there
* the code is cleaner
* no extra variables are created

Sample policy:

```
bundle agent main
{
  vars:
      "parsed" data => data_regextract("^(?<name1>...)(...)(..)-(?<name2>...)-(..).*", "abcdef12-345-67andsoon");
      "parsed_str" string => format("%S", parsed);

  reports:
      "$(this.bundle): '$(parsed[0])' parses into: $(parsed_str)";
}
```

Output:

```console
% cf-agent -KI -f ~/Dropbox/cf/test/test_data_regextract.cf
R: main: 'abcdef12-345-67andsoon' parses into: {"0":"abcdef12-345-67andsoon","name1":"abc","2":"def","3":"12","name2":"345","5":"67"}
```

I'll finish this with docs and tests if it's acceptable.  I think it's very useful.

### CODE REVIEW

**Comment by nickanderson** [comment]
i like it!

Named captures is really nice, and anything that avoids the 4k limit is handy

**Comment by tzz** [comment]
ping @kacfengine @jimis @estenberg ?

**Comment by estenberg** [comment]
Really nice! Go for it. :+1:

**Comment by basvandervlies** [comment]
+1

**Comment by tzz** [comment]
OK, tests and example added.  Documentation is in https://github.com/cfengine/documentation/pull/1008

**Comment by tzz** [comment]
ping?

File: libutils/regex.c [hunk 1]
```diff
@@ -172,12 +197,43 @@ Seq *StringMatchCaptures(const char *regex, const char *str)
...
+                {
+                    capture = BufferNewFrom(tabptr + 2, name_entry_size - 3);
+                    break;
+                }
+                tabptr += name_entry_size;
+            }
+        }
+
+        if (return_names)
+        {
```
Comment (kacfengine) [inline libutils/regex.c]: This is guaranteed to be true if ``have_named_captures`` is true (it depends on it), so just move this block into the previous block.

File: libutils/regex.c [hunk 2]
```diff
@@ -172,12 +197,43 @@ Seq *StringMatchCaptures(const char *regex, const char *str)
...
-                                ovector[2*i + 1] - ovector[2 * i]));
+        Buffer *capture = NULL;
+
+        if (have_named_captures)
+        {
+            // The overhead of doing a nested name scan is negligible.
+            tabptr = name_table;
+            for (int namepos = 0; namepos < namecount; namepos++)
+            {
+                int n = (tabptr[0] << 8) | tabptr[1];
```
Comment (kacfengine) [inline libutils/regex.c]: Ugh, the PCRE API really sucks in some aspects! :-P

File: libutils/regex.c [hunk 3]
```diff
@@ -133,7 +135,14 @@ bool StringMatchFullWithPrecompiledRegex(pcre *pattern, const char *str)
     }
 }
 
-Seq *StringMatchCaptures(const char *regex, const char *str)
+// Returns a Sequence with Buffer elements.
+
+// If return_names is set, for every capture there will be the name or
+// number of the capturing group, followed by the captured data.
```
Comment (kacfengine) [inline libutils/regex.c]: Suggest slight clarification that even numbered indexes will be name or number, and odd numbered will be capture data. At first I thought you meant returning some kind of pair structure.

**Comment by kacfengine** [comment]
Looks good otherwise, the named captures will be a real boon I think. :-)

**Comment by tzz** [comment]
I updated the mergedata.cf example and the documentation is in https://github.com/cfengine/documentation/pull/1011

File: libutils/regex.c (hunk 1 above)
Comment (tzz) [inline libutils/regex.c]: OK; done.

File: libutils/regex.c (hunk 3 above)
Comment (tzz) [inline libutils/regex.c]: OK; done

**Comment by tzz** [comment]
Adjusted as requested, thank you very much for reviewing this.

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
During ``make check`` I got:

<pre>
11:57:35 ./.libs/libstr.a(regex.o): In function `StringMatchCaptures':
11:57:35 /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_x86_64_linux_redhat_6/core/tests/unit/../../libutils/regex.c:203: undefined reference to `BufferDestroy'
11:57:35 /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_x86_64_linux_redhat_6/core/tests/unit/../../libutils/regex.c:217: undefined reference to `BufferNewFrom'
11:57:35 /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_x86_64_linux_redhat_6/core/tests/unit/../../libutils/regex.c:225: undefined reference to `BufferNew'
11:57:35 /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_x86_64_linux_redhat_6/core/tests/unit/../../libutils/regex.c:226: undefined reference to `BufferAppendF'
11:57:35 /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_x86_64_linux_redhat_6/core/tests/unit/../../libutils/regex.c:232: undefined reference to `BufferNewFrom'
11:57:35 collect2: ld returned 1 exit status
</pre>

**Comment by tzz** [comment]
@kacfengine I'm not sure what's up, it works for me on Ubuntu 14.10 with gcc.

**Comment by cdituri** [comment]
tests/unit/Makefile.am:~L63, add `../../libutils/buffer.c` maybe? (not tested, just a guess)
```
check_LTLIBRARIES += libstr.la
libstr_la_SOURCES = ../../libutils/string_lib.c ../../libutils/regex.c \
	../../libutils/encode.c ../../libutils/writer.c \
	../../libutils/sequence.c
libstr_la_LIBADD = libtest.la
```

**Comment by tzz** [comment]
ping

**Comment by kacfengine** [comment]
I tried fixing this quickly like cdituri suggested, but it quickly became non-trivial. Are you saying you can run ``make check`` successfully on all of core? I'm using Ubuntu 12.04.

**Comment by tzz** [comment]
No, sorry, I was wrong.  It fails for me too.  But I don't know the build system well enough to fix it; it seems like it's a problem outside of the scope of this change.

**Comment by kacfengine** [comment]
I was able to fix the build system problem, you can cherry-pick this commit to your branch: https://github.com/kacfengine/core/commit/620a0ae1a3535fbcc2af5dbcb9f0d7b345e71b1a.

However, now I get a bunch of failures in acceptance tests, for example in ``00_basics/02_switches/002.cf``. At first glance it looks like a failed expansion:

<pre>
notice: Q: ".../cf-agent" -Kf ": R: $(fn[1]) Pass
R: Running: "/root/code/core/tests/acceptance/workdir/__00_basics_02_switches_002_cf/bin/cf-agent" -Kf /root/code/core/tests/acceptance/./00_basics/02_switches/002.cf.sub -D AUTO,ok,
R: Test results come from /root/code/core/tests/acceptance/./00_basics/02_switches/002.cf.sub

Return code is 0.

  ==> FAIL (UNEXPECTED FAILURE)
</pre>

**Comment by tzz** [comment]
I've rebased and the compilation works, but I get segfaults when I test any regextract functions.  I'll debug.

**Comment by tzz** [comment]
OK, fixed.  `have_named_captures` didn't imply `return_names`.

**Comment by kacfengine** [comment]
Ah right, ``have_named_captures`` means ``return_names`` is true, but not the other way around. Yes, you're right, my optimization is bogus then.

**Comment by kacfengine** [comment]
trigger build

File: tests/unit/Makefile.am [hunk 4]
```diff
@@ -321,8 +330,11 @@ connection_management_test_SOURCES = connection_management_test.c ../../cf-serve
 connection_management_test_LDADD = ../../libpromises/libpromises.la libtest.la ../../cf-serverd/libcf-serverd.la
 
 rlist_test_SOURCES = rlist_test.c \
-       ../../libpromises/rlist.c ../../libutils/logging.c
-rlist_test_LDADD = libtest.la libstr.la ../../libpromises/libpromises.la
+	../../libpromises/rlist.c \
+rlist_test_LDADD = libtest.la ../../libpromises/libpromises.la
```
Comment (kacfengine) [inline tests/unit/Makefile.am]: Oops, my bad, I left a stray ``\`` after the ``rlist.c`` entry there. It's easier if you fix it on your branch so I don't have to make a new PR.

**Comment by tzz** [comment]
OK, amended and pushed with `-f`.

**Comment by tzz** [comment]
trigger build

**Comment by tzz** [comment]
(I don't know if that will work, but what the hey)

**Comment by kacfengine** [comment]
No, unfortunately not. :-)

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
Still getting errors in ``04_examples/outputs/check_outputs.cf``:

<pre>
Q: "...r/bin/perl ../.": Command: "/home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_i386_linux_debian_4/core/tests/acceptance/workdir/__04_examples_outputs_check_outputs_cf/bin/cf-agent" -D_cfe_output_testing -nKf /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_i386_linux_debian_4/core/tests/acceptance/workdir/__04_examples_outputs_check_outputs_cf/tmp/TEST.cfengine/mergedata.cf 2>&1
Q: "...r/bin/perl ../.": NEW OUTPUT: [[[R: merging {"a":[1,2,3],"b":[]} with {"b":[4,5,6]} produced {"a":[1,2,3],"b":[4,5,6]}
Q: "...r/bin/perl ../.": R: $(merged_d1_wrap_array_d2_str)
Q: "...r/bin/perl ../.": R: $(merged_d1_wrap_map_d2_str)
Q: "...r/bin/perl ../.": R: merging {"a":[1,2,3],"b":[]} with [4,5,6] produced {"a":[1,2,3],"b":[],"0":4,"1":5,"2":6}
Q: "...r/bin/perl ../.": R: merging [4,5,6] with { "element1", "element2" } produced [4,5,6,"element1","element2"]
Q: "...r/bin/perl ../.": R: merging {"a":[1,2,3],"b":[]} with array1 produced {"a":[1,2,3],"b":[],"mykey":["array_element1","array_element2"]}
Q: "...r/bin/perl ../.": R: merging {"b":[4,5,6]} with array2 produced {"b":[4,5,6],"otherkey":"hello"}
Q: "...r/bin/perl ../.": R: test2: merged containers with cmerge = {"a":"1","b":"2","c":"3","d":"4"}
Q: "...r/bin/perl ../.": ]]]
Q: "...r/bin/perl ../.": --- /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_i386_linux_debian_4/core/tests/acceptance/workdir/__04_examples_outputs_check_outputs_cf/tmp/TEST.cfengine/mergedata.cf.a	2015-04-17 10:23:06.000000000 +0200
Q: "...r/bin/perl ../.": +++ /home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_i386_linux_debian_4/core/tests/acceptance/workdir/__04_examples_outputs_check_outputs_cf/tmp/TEST.cfengine/mergedata.cf.b	2015-04-17 10:23:06.000000000 +0200
Q: "...r/bin/perl ../.": @@ -1,6 +1,6 @@
Q: "...r/bin/perl ../.":  R: merging {"a":[1,2,3],"b":[]} with {"b":[4,5,6]} produced {"a":[1,2,3],"b":[4,5,6]}
Q: "...r/bin/perl ../.": -R: merging {"a":[1,2,3],"b":[]} with wrapped [ {"b":[4,5,6]} ] produced {"a":[1,2,3],"b":[],"0":{"b":[4,5,6]}}
Q: "...r/bin/perl ../.": -R: merging {"a":[1,2,3],"b":[]} with wrapped { "newkey": {"b":[4,5,6]} produced {"a":[1,2,3],"b":[],"newkey":{"b":[4,5,6]}}
Q: "...r/bin/perl ../.": +R: $(merged_d1_wrap_array_d2_str)
Q: "...r/bin/perl ../.": +R: $(merged_d1_wrap_map_d2_str)
Q: "...r/bin/perl ../.":  R: merging {"a":[1,2,3],"b":[]} with [4,5,6] produced {"a":[1,2,3],"b":[],"0":4,"1":5,"2":6}
Q: "...r/bin/perl ../.":  R: merging [4,5,6] with { "element1", "element2" } produced [4,5,6,"element1","element2"]
Q: "...r/bin/perl ../.":  R: merging {"a":[1,2,3],"b":[]} with array1 produced {"a":[1,2,3],"b":[],"mykey":["array_element1","array_element2"]}
Q: "...r/bin/perl ../.": ../../examples/mergedata.cf: output differs from original...
error: Method "verbose_output" failed in some repairs
</pre>

While I could probably easily have fixed that myself, the slightly more disturbing thing is that I get this on Debian 4 32-bit (and only there):

<pre>
Error Message

FAIL (UNEXPECTED FAILURE) ./01_vars/02_functions/mergedata.cf

Stacktrace

lt-cf-agent: var_expressions.c:314: VarRefParseFromNamespaceAndScope: Assertion `ns == ((void *)0) && "A variable missing a scope should not have a namespace"" failed.
/home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_i386_linux_debian_4/core/tests/acceptance/workdir/__01_vars_02_functions_mergedata_cf/runtest: line 8:  6172 Aborted                 "/home/jenkins/workspace/testing-enterprise-pr_core/label/PACKAGES_i386_linux_debian_4/core/tests/acceptance/../../cf-agent/cf-agent" -Kf "./01_vars/02_functions/mergedata.cf" -D AUTO,DEBUG
</pre>

Sorry to give it back to you, but I don't have time to look at deep code issues like these.

**Comment by tzz** [comment]
AHA!  The bug was that `StringMatchCaptures` now returns `Buffer` not a `char*`.  Fixed.

**Comment by tzz** [comment]
I explicitly checked the source tree to ensure there was no other place where `StringMatchCaptures` might have been used in the old way, expecting `char*`.  Thanks for catching this one.

**Comment by tzz** [comment]
This bug happened because the wrap map/array functionality wasn't in the code when I wrote the PR originally :)  It's obvious now...

**Comment by kacfengine** [comment]
Aha, inter-branch bugs, the one thing the pull request analyzer cannot catch. :-)

**Comment by kacfengine** [comment]
trigger build

**Comment by kacfengine** [comment]
Merged! Thanks for fixing!

**Pull request merged**